platforms/nuttx/src/px4/common/srgbled/srgbled.cpp

```diff
@@ -0,0 +1,166 @@
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+* @file srgbled.cpp
+* Author: David.Sidrane@Nscdg.com*
+*  This is a driver for the the neopixel class of serial RGB LEDs.
+*
+*  It is arch dependent on ARM for the DWT and the GPIO is done in arch
+*  independent manner.
+*
+*  To use this driver the board must define:
+
+*  BOARD_HAS_N_S_RGB_LED  - the number of LEDs
+*  BOARD_SRGBLED_BIT      - the bit number it is connected to. 0-n (not a mask)
+        and
+*  BOARD_SRGBLED_PORT     - The address of the port's "data out" register
+*                           the LED is connected to.
+*       OR
+*  BOARD_SRGBLED_SET_PORT - for an arch with set/clear GPIO. The address of the
+*                           port's "SET Bit" register
+*  BOARD_SRGBLED_CLEAR    - for an arch with set/clear GPIO.  The address of the
+*                           port's "Clear Bit" register
+*
+*  N.B. This version is small but will disable interrupts for 30 us per LED!
+*  With an 8 LED's this is 240uS and not tolerable for systems with sensors.
+*  It is however useful in bootloaders.
+*
+*  The DMA version of this driver should be used for systems that can not have
+*  interrupts for long times.
+*
+*/
+
+#include <px4_platform_common/px4_config.h>
+#include <board_config.h>
+#include <dwt.h>
+#include <nvic.h>
+
+#define REG(_addr)       (*(volatile uint32_t *)(_addr))
+#define rDEMCR           REG(NVIC_DEMCR)
+#define rDWT_CTRL        REG(DWT_CTRL)
+#define rDWT_CNT         REG(DWT_CYCCNT)
+
+#if defined(BOARD_SRGBLED_PORT) && defined(BOARD_SRGBLED_BIT)
+#  define PORT             REG(BOARD_SRGBLED_PORT)
+#  define D0               ((PORT) &= ~(1 << BOARD_SRGBLED_BIT));
+#  define D1               ((PORT) |= (1 << BOARD_SRGBLED_BIT));
+#elif defined(BOARD_SRGBLED_SET_PORT) && defined(BOARD_SRGBLED_CLEAR_PORT)  && defined(BOARD_SRGBLED_BIT)
+#  define PORT             REG(BOARD_SRGBLED_PORT)
+#  define D0               ((BOARD_SRGBLED_CLEAR_PORT) |= (1 << BOARD_SRGBLED_BIT));
+#  define D1               ((BOARD_SRGBLED_SET_PORT)   |= (1 << BOARD_SRGBLED_BIT));
+#else
+# error BOARD_SRGBLED_[]{SET|CLEAR}_]PORT and BOARD_SRGBLED_BIT needs to be defined.
+#endif
+
+#define DWT_DEADLINE(t)  rDWT_CNT + (t)
+#define DWT_WAIT(v, D)   while((rDWT_CNT - (v)) < (D)){}
+
+#define T0H              (STM32_SYSCLK_FREQUENCY/3333333)
+#define T1H              (STM32_SYSCLK_FREQUENCY/1666666)
+#define TW               (STM32_SYSCLK_FREQUENCY/850000)
+
+typedef union {
+	uint8_t  grb[3];
+	uint32_t l;
+} led_data_t;
+
+#if defined(BOARD_HAS_N_S_RGB_LED) && !defined(S_RGB_LED_DMA)
+
+
+int led_pwm_servo_set(unsigned channel, uint8_t  value);
```

Comment:
I'm not sure I follow. You implement empty stubs for the led_pwm API, then do all the actual work in a new function (led_pwm_servo_set_ex) that isn't part of that API? Why not just create a new frontend driver for the neopixel?

Reply:
Yes. I pushed so you could get your part working. 
This was done just to test the new physical driver code (scopeing the data) with working RGB code. The NeoPixel driver should have an addressable LED interface (as supported in the LED API) and the low level will have to support that as well. But this should get us started.